### PR TITLE
PCHR-3870: Provide Functionality for getting Leave Ical Feed Via URL

### PR DESCRIPTION
**Pull request by tunbola**
## Overview
This PR adds functionality for actual Leave request Ical feed generation off the Feed configuration.
This PR also depends on https://github.com/compucorp/civihr-employee-portal/pull/511 which allows the calendar feed URL not to be blocked by the public firewall and to be publicly accessible. 

## Before
- The functionality for generating Leave request Ical feed was not present.
## After
Trying to retrieve the Ical for a disabled feed or non existent hash

feedgenerator2

Trying to retrieve the Ical for a valid and enabled feed

feedgenerator

## Technical Details
- A new public URL `civicrm/calendar-feed` was added that allows a user add a calendar feed to his preferred calendar application. 
A fully formed feed URL is similar to: `civicrm/calendar-feed?hash=7f9e70bebdc30dab872829893c47a3ac` where the hash is the feed config hash.
- A new class `LeaveRequestCalendarFeedData` was added to encapsulate the logic needed for fetching associated data for a feed configuration. The class also provides a get method for returning prepared data in a format expected for the Leave request data Ical conversion.
- A `LeaveRequestCalendarFeedIcal` was also added that accepts an `LeaveRequestCalendarFeedData` object in its get method and converts the prepared data to an Ical format.
- For the purpose of display in the calendar, some changes were made to time for Leave requests in days:
 AM - 6 AM to 12 PM
 PM - 12:01 PM to 7 PM
 All Day - 6 AM to 7 PM

A new `LeaveTime` CalendarFeed helper class was added to adjust the dates.
- The calendar feed URL when visited for a valid calendar feed downloads an ical file as an attachment.

A PHP Ical library https://github.com/zcontent/icalendar, was used in generating the Ical format off the Leave request calendar feed data. The library itself was added to a vendor folder in the HRCore extension.
The calendar feed URL invokes the get method of the `CRM_HRLeaveAndAbsences_Page_LeaveRequestCalendarFeed` function
```php
  <item>
    <path>civicrm/calendar-feed</path>
    <page_callback>CRM_HRLeaveAndAbsences_Page_LeaveRequestCalendarFeed::get</page_callback>
    <title>LeaveCalendarFeed</title>
    <access_callback>1</access_callback>
    <is_public>true</is_public>
  </item>
```
The get method is used to render the Leave Ical data as a browser attachment
```php
  public static function get() {
    try {
      $feedHash = CRM_Utils_Array::value('hash', $_GET);
      $leaveFeedData = new LeaveRequestCalendarFeedData($feedHash);
      $leaveFeedIcal = new LeaveRequestCalendarFeedIcal();
      $leaveFeedIcal = $leaveFeedIcal->get($leaveFeedData);

      header('Content-Type: text/calendar; charset=utf-8');
      header('Content-Disposition: attachment; filename="cal.ics"');
      echo $leaveFeedIcal;
      die();

    } catch(Exception $e) {
      http_response_code(404);
      echo $e->getMessage();
      die();
    }
  }
```

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedDataTest.php [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+      'contact_id' => $contact1['id'],
+      'first_name' => $contact1['first_name'],
+      'last_name' => $contact1['last_name'],
+      'type_id' => $absenceType2->id,
+      'from_date' => CRM_Utils_Date::processDate('today'),
+      'to_date' => CRM_Utils_Date::processDate('+2 days'),
+    ];
+
+    //disable absence type 2
+    AbsenceTypeFabricator::fabricate(['id' => $absenceType2->id, 'is_active' => 0]);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedDataTest.php]: Why don't you create it as disabled? Do we really need this separate statement to make it disabled?

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedDataTest.php [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    $feedConfig1 = LeaveCalendarFeedConfigFabricator::fabricate([
+      'title' => 'Feed 1',
+      'composed_of' => [
+        'leave_type' => [$absenceType->id, $absenceType2->id],
+      ]
+    ]);
+
+    $leaveFeedData = new LeaveRequestCalendarFeedData($feedConfig1->hash);
+
+    unset($params[2]);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedDataTest.php]: Document why you are unsetting this

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedDataTest.php [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+    $feedConfig1 = LeaveCalendarFeedConfigFabricator::fabricate([
+      'title' => 'Feed 1',
+      'composed_of' => [
+        'leave_type' => [$absenceType->id],
+      ]
+    ]);
+
+    $leaveFeedData = new LeaveRequestCalendarFeedData($feedConfig1->hash);
+
+    unset($params[1], $params[2]);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedDataTest.php]: Document why you're unsetting this

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedIcal.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    foreach ($feedData->get() as $data) {
+      $eventObject = new ZCiCalNode("VEVENT", $icalObject->curnode);
+      $eventObject->addNode(new ZCiCalDataNode("SUMMARY:" . $data['display_name']));
+      $eventObject->addNode(new ZCiCalDataNode("DTSTART:" . ZCiCal::fromSqlDateTime($data['from_date'])));
+      $eventObject->addNode(new ZCiCalDataNode("DTEND:" . ZCiCal::fromSqlDateTime($data['to_date'])));
+
+      // We need to create a unique string for this event so that on reimport by calendar app
+      // Event is not duplicated and changes will be updated.
+      $uid = $data['id'] . '_feed_' . $data['contact_id'];
```
[the remark this answers is not in the document]
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedIcal.php]: No reason really. It should be enough

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Page/LeaveRequestCalendarFeed.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    try {
+      $feedHash = CRM_Utils_Array::value('hash', $_GET);
+      $leaveFeedData = new LeaveRequestCalendarFeedData($feedHash);
+      $leaveFeedIcal = new LeaveRequestCalendarFeedIcal();
+      $leaveFeedIcal = $leaveFeedIcal->get($leaveFeedData);
+
+      header('Content-Type: text/calendar; charset=utf-8');
+      header('Content-Disposition: attachment; filename="cal.ics"');
+      echo $leaveFeedIcal;
+      die();
```
[the remark this answers is not in the document]
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Page/LeaveRequestCalendarFeed.php]: Never knew a function like this existed!

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedData.php [hunk 6]
```diff
@@ -87,6 +94,21 @@ public function get() {
     return $leaveData;
   }
 
+  /**
+   * Returns the Leave type label from the leave type ID. For cases
+   * where the typeId is not found in the absence types list, i.e for
+   * leave types with their hide_label property as true, the leave label
+   * is returned as 'Leave'.
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedData.php]: I'm not sure if the `hide_label` makes any difference here, as the `getEnabledLeaveTypes()` only return a list of enabled absence types, regardless if the label should be hidden or not. I'm missing something?

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequestCalendarFeedConfig.php [hunk 7]
```diff
@@ -430,4 +431,62 @@ public function getInaccessibleFeedConfigs(
...
+    return $feedConfig;
+  }
+
+  /**
+   * Returns the departments the feed config has set for the
+   * composed of property.
+   *
+   * @return string
+   */
+  public function getDepartmentsComposedOf() {
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequestCalendarFeedConfig.php]: Do we really need the `ComposedOf` as part of the method name? Why not simply `getDepartments()`, `getLocations()` etc? It doesn't matter for the other classes that this information is stored as an array inside `composed_of`

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedIcal.php [hunk 8]
```diff
@@ -18,27 +21,26 @@ class CRM_HRLeaveAndAbsences_Service_LeaveRequestCalendarFeedIcal {
...
 
-      // We need to create a unique string for this event so that on reimport by calendar app
-      // Event is not duplicated and changes will be updated.
-      $uid = $data['id'] . '_feed_' . $data['contact_id'];
-      $eventObject->addNode(new ZCiCalDataNode("UID:" . $uid));
-      $eventObject->addNode(new ZCiCalDataNode("DTSTAMP:" . ZCiCal::fromSqlDateTime()));
+      // The leave request is used as a string for this event so that on reimport by calendar app
+      // the event is not duplicated and changes will be updated.
+      $eventObject->addNode(new ZCiCalDataNode("UID:" . $data['id']));
+      $eventObject->addNode(new ZCiCalDataNode("DTSTAMP:" . ZCiCal::fromSqlDateTime($dateTime)));
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedIcal.php]: Why was this change (the $dateTime) necessary? If it's only to help with the test, then we need to find another way to test it. We cannot add new properties to classes only because we need it for a test

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedData.php [hunk 9]
```diff
@@ -109,13 +117,24 @@ private function getLeaveTypeLabel($typeId) {
...
+   * Returns the datetime for this leave feed data class
+   * was instantiated.
    *
-   * @return LeaveRequestCalendarFeedConfig
+   * @return \DateTime
    */
-  public function getFeedConfig() {
-    return $this->feedConfig;
+  public function getInstantiatedDateTime() {
+    return $this->instantiatedDateTime;
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedData.php]: Why was this change necessary? If it's only to help with the test, then we need to find another way to test it. We cannot add new properties to classes only because we need it for a test

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedData.php [hunk 10]
```diff
@@ -68,12 +74,13 @@ public function get() {
...
+
 
     $leaveData = [];
     foreach($leaveRequests as $leaveRequest) {
-      $leaveTypeName =
-        !empty($absenceTypesList[$leaveRequest['type_id']]) ? $absenceTypesList[$leaveRequest['type_id']] : 'Leave';
+      $leaveTypeName = $this->getLeaveTypeLabel($leaveRequest['type_id']);
+      //We need to adjust the time for leave request in days for display in the calendar
+      //different from the default time stored in the leave request table.
       CalendarLeaveTimeHelper::adjust($leaveRequest);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedData.php]: It is a little bit better now, but the comment says what is this for, but not why we need to adjust things. Please explain why.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Helper/CalendarFeed/LeaveTime.php [hunk 11]
```diff
@@ -4,6 +4,9 @@
 
 /**
  * Class CRM_HRLeaveAndAbsences_Helper_CalendarFeed_LeaveTime
+ *
+ * This is an helper class that helps to modify the time for leave request
+ * in days for display in a calendar.
  */
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Helper/CalendarFeed/LeaveTime.php]: Please also describe why we need this and what kinds of adjustment it does

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Page/LeaveRequestCalendarFeed.php [hunk 12]
```diff
@@ -19,15 +19,12 @@ public static function get() {
       $leaveFeedIcal = new LeaveRequestCalendarFeedIcal();
       $leaveFeedIcal = $leaveFeedIcal->get($leaveFeedData);
 
-      header('Content-Type: text/calendar; charset=utf-8');
-      header('Content-Disposition: attachment; filename="cal.ics"');
-      echo $leaveFeedIcal;
-      die();
+      CRM_Utils_System::download('cal.ics', 'text/calendar', $leaveFeedIcal);
+      CRM_Utils_System::civiExit();
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Page/LeaveRequestCalendarFeed.php]: `civiExit()` is called by `download()` automatically (https://github.com/civicrm/civicrm-core/blob/fee14197b427c1781e369e5bfd36816afad6d7ee/CRM/Utils/System.php#L849-L852), so there's no need to call it here yourself

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedData.php (hunk 6 above)
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequestCalendarFeedData.php]: Yeah, thats true.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequestCalendarFeedConfig.php (hunk 7 above)
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequestCalendarFeedConfig.php]: There are departments and locations also stored in the visible_to property. So we need to distinguish between the departments the feed is composed of and the departments the feed link is visible to.
